common/app/components/SubtitlePlayer.tsx

```diff
@@ -511,6 +538,75 @@ export default function SubtitlePlayer({
         };
     }, []);
 
+    const subtitleColoringRef = useRef<SubtitleColoring | undefined>(undefined);
+
+    useEffect(() => {
+        if (channel) return; // Handled by extension or VideoPlayer
+
+        const subtitleColoring = new SubtitleColoring(
+            settings,
+            (updatedSubtitles) => {
+                const newSubtitles = updateSubtitleColors(subtitleListRef.current, updatedSubtitles);
+                if (newSubtitles) onSubtitles(newSubtitles);
+            },
+            () => clockRef.current.time(lengthRef.current)
+        );
+        if (subtitleListRef.current) subtitleColoring.subtitles = subtitleListRef.current;
+        subtitleColoring.bind();
+        subtitleColoringRef.current = subtitleColoring;
+
+        return () => {
+            subtitleColoringRef.current?.unbind();
+            subtitleColoringRef.current = undefined;
+        };
+    }, [channel, settings, onSubtitles]);
+
+    useEffect(() => {
+        if (!subtitleColoringRef.current) return;
+        subtitleColoringRef.current.subtitles = subtitleListRef.current ?? [];
+    }, [subtitles]);
+
+    // Immediate update of subtitle colors when changed (from extension/VideoPlayer)
+    useEffect(() => {
+        return channel?.onSubtitlesUpdated((updatedSubtitles) => {
+            const newSubtitles = updateSubtitleColors(subtitleListRef.current, updatedSubtitles);
+            if (newSubtitles) onSubtitles(newSubtitles);
+        });
+    }, [channel, onSubtitles]);
+
+    // Poll to catch ones that slip if request fails
+    useEffect(() => {
+        if (!channel && !subtitleColoringRef.current) return;
+
+        const poll = async () => {
+            const subtitleColoring = subtitleColoringRef.current; // If not from extension/VideoPlayer
+            if (subtitleColoring) {
+                const newSubtitles = updateSubtitleColors(subtitleListRef.current, subtitleColoring.subtitles);
+                if (newSubtitles) onSubtitles(newSubtitles);
+            } else if (channel) {
+                if (!channel.fromExtension) {
```

Comment:
I don't think this case is necessary if `SubtitlePlayer` is publishing updated subtitles already. As a result the new field `fromExtension` can be removed.